src/storage/mongodb/SensitiveDataMigrationStorage.ts

```diff
@@ -0,0 +1,87 @@
+import { SensitiveDataMigrationState, SettingSensitiveData } from '../../types/SensitiveData';
+
+import BackendError from '../../exception/BackendError';
+import Constants from '../../utils/Constants';
+import Cypher from '../../utils/Cypher';
+import DatabaseUtils from './DatabaseUtils';
+import { ObjectID } from 'mongodb';
+import Utils from '../../utils/Utils';
+import global from './../../types/GlobalType';
+
+const MODULE_NAME = 'SensitiveDataMigrationStorage';
+
+export default class SensitiveDataMigrationStorage {
+  private static migrationId: ObjectID;
```

Comment:
It's not really a migration task like the others. It do not used the current migration framework made to run migration only at startup. 
It's a scheduler task that use it's own collection to track the migration state that permit to survive to a migration stopped in the middle.

The id is required to have a reference to the current migration run but it should follow the same naming used everywhere: id in the object and _id in the DB.

Reply:
If there is a dedicated collection, the write in the 'settings' and in 'sensitivedatamigrations' must be done in a transaction otherwise you can manage writing the settings and fail writing the sensitivedatamigrations.
I think tracking the changes only in settings collection should sufficient and would spare the code to handle the sensitivedatamigrations.

Reply:
> If there is a dedicated collection, the write in the 'settings' and in 'sensitivedatamigrations' must be done in a transaction otherwise you can manage writing the settings and fail writing the sensitivedatamigrations.

Both actions can be considered independents: 
+ in one collection, you only store the settings
+ in other collection, you keep track of the migration statuses in a transactional manner (at at the beginning: self contained data structure to restart a transaction for debugging)   

> I think tracking the changes only in settings collection should sufficient and would spare the code to handle the sensitivedatamigrations.

It can help to have a separate collection to permit to keep a migration history per tenant for debug purpose. I've just took the safe design road on that migration. I don't mind to only have one record history in settings but since the code is already there, not sure it worth it. 

You've scheduled a meeting with Stefania. So I assume you took over its development framing.

Reply:
Even if the code is already there, it doesn't mean we have to keep it only for that reason.
As usual, I'm the final reviewer but I need your feedback on the choices that have been taken which Stefania knows I think, thus the meeting.
I'll get back to you to keep you in sync.

Reply:
> Even if the code is already there, it doesn't mean we have to keep it only for that reason.

The main reason is the separation of the data used by different process: one for the cyphering settings handling and one for the migration states. The settings determine if a migration shall be done and then the migration process starts and write its own data separately. The historisation is a design bonus. As well as not keeping sensitive informations ciphered strings close to their keys and settings once the migration states are expunged from the transactional fields used for debug. The deletion of the migration states was planned at the beginning once the migration was successful to permit to keep the transactional extra fields in case of errors or issue while not opening severe security issues as a tradeoff.    

I think you have the informations needed.